src/nvim/change.c

```diff
@@ -523,6 +503,13 @@ void deleted_lines_mark(linenr_T lnum, int count)
 /// @param xtra number of extra lines (negative when deleting)
 void changed_lines_redraw_buf(buf_T *buf, linenr_T lnum, linenr_T lnume, linenr_T xtra)
 {
+  // When lines are deleted "lnume" needs to be adjusted to ensure it
+  // is set to the last line below the change.
+  lnume += (xtra < 0);
```

Comment:
This doesn't look correct. `lnume` is already the line below last change line as mentioned in the doc comment above.

Reply:
It didn't look right to me either but "it" in the comment should be `buf->b_mod_bot`. You'll find that when you delete a line 1, currently both `buf->b_mod_top` _and_ `buf->b_mod_bot` are set to 1. That is wrong according to
https://github.com/neovim/neovim/blob/c73a827564083b4dadb3a70468466306fb416d08/src/nvim/drawscreen.c#L1574

Commenting out this line results in a failure `decorations_spec.lua`.

Reply:
How is that wrong? The lines that need changes have been deleted, so no lines need changes.

Reply:
It's wrong because it results in missing lines. Maybe `mod_top` shouldn't be set either then?

Reply:
The lines that are deleted by the scroll logic in #33358.

Reply:
But yeah I missed, or forgot about #27453. If the accuracy of that approach is preferred and you intend to pick that back up we can close this or reduce it to just the refactor part.

Reply:
> so the condition here should be related to extmarks

But since left gravity marks can move from the line beyond the change as you found in #27369, I think this is basically needed for any kind of decoration. The most restrictive condition we have available for that right now would be something like `buf->b_marktree->n_keys > 0`.